cumulus/client/bootnodes/src/advertisement.rs

```diff
@@ -0,0 +1,592 @@
+			}
+		} else {
+			// First advertisement on startup.
+			let (current_epoch_key, next_epoch_key) =
+				self.current_and_next_epoch_keys(header).await;
+			self.current_epoch_key = current_epoch_key.clone();
+			self.next_epoch_key = next_epoch_key.clone();
+
+			if let Some(current_epoch_key) = current_epoch_key {
+				debug!(
+					target: LOG_TARGET,
+					"Initial advertisement of bootnode for current epoch key {}",
+					hex::encode(current_epoch_key.as_ref()),
+				);
+
+				self.relay_chain_network.start_providing(current_epoch_key);
+			} else {
+				warn!(
+					target: LOG_TARGET,
+					"Initial advertisement of bootnode for current epoch failed: no key."
+				);
+			}
+
+			if let Some(next_epoch_key) = next_epoch_key {
+				debug!(
+					target: LOG_TARGET,
+					"Initial advertisement of bootnode for next epoch key {}",
+					hex::encode(next_epoch_key.as_ref()),
+				);
+
+				self.relay_chain_network.start_providing(next_epoch_key);
+			} else {
+				warn!(
+					target: LOG_TARGET,
+					"Initial advertisement of bootnode for next epoch failed: no key."
+				);
+			}
+		}
+	}
+
+	/// The list of parachain side addresses.
+	///
+	/// The addresses are sorted as follows:
+	///  1) public addresses provided by the operator
+	///  2) global listen addresses
+	///  3) discovered external addresses
+	///  4) non-global listen addresses
+	///  5) loopback listen addresses
+	fn paranode_addresses(&self) -> Vec<Multiaddr> {
+		let local_peer_id = self.parachain_network.local_peer_id();
+
+		// Discard `/p2p/<peer_id>` part. `None` if the address contains foreign peer ID.
+		let without_p2p = |mut addr: Multiaddr| match addr.iter().last() {
+			Some(Protocol::P2p(peer_id)) if &peer_id == local_peer_id.as_ref() => {
+				addr.pop();
+				Some(addr)
+			},
+			Some(Protocol::P2p(_)) => {
+				warn!(
+					target: LOG_TARGET,
+					"Ignoring parachain side address containing not our peer ID: {addr}",
+				);
+				None
+			},
+			_ => Some(addr),
+		};
+
+		// Check if the address is global.
+		let is_global = |address: &Multiaddr| {
+			address.iter().all(|protocol| match protocol {
+				// The `ip_network` library is used because its `is_global()` method is stable,
+				// while `is_global()` in the standard library currently isn't.
+				Protocol::Ip4(ip) => IpNetwork::from(ip).is_global(),
+				Protocol::Ip6(ip) => IpNetwork::from(ip).is_global(),
+				_ => true,
+			})
+		};
+
+		// Check if the address is a loopback address.
+		let is_loopback = |address: &Multiaddr| {
+			address.iter().any(|protocol| match protocol {
+				Protocol::Ip4(ip) => IpNetwork::from(ip).is_loopback(),
+				Protocol::Ip6(ip) => IpNetwork::from(ip).is_loopback(),
+				_ => false,
+			})
+		};
+
+		// 1) public addresses provided by the operator
+		let public_addresses = self.public_addresses.clone().into_iter();
+
+		// 2) global listen addresses
+		let global_listen_addresses =
+			self.parachain_network.listen_addresses().into_iter().filter(is_global);
```

Comment:
nit; should `is_global` take into account the provided `--allow-private-ip`/similar flags?

Reply:
Yes, this is done below by checking `self.advertise_non_global_ips`.